substrate/frame/election-provider-multi-block/src/signed/mod.rs

```diff
@@ -163,44 +164,95 @@ impl<T: Config> SolutionDataProvider for Pallet<T> {
 						Precision::BestEffort,
 					);
 					debug_assert!(_res.is_ok());
+				} else {
+					// No leader to reward; nothing to do.
+					sublog!(
+						warn,
+						"signed",
+						"Tried to report Queued but no leader was present for round {}",
+						current_round
+					);
+					defensive!("should never happen");
 				}
 			},
 			VerificationResult::Rejected => {
-				// defensive: if there is a result to be reported, then we must have had some
-				// leader.
-				if let Some((loser, metadata)) =
-					Submissions::<T>::take_leader_with_data(Self::current_round()).defensive()
-				{
-					// first, let's slash their deposit.
-					let slash = metadata.deposit;
-					let _res = T::Currency::burn_held(
-						&HoldReason::SignedSubmission.into(),
-						&loser,
-						slash,
-						Precision::BestEffort,
-						Fortitude::Force,
-					);
-					debug_assert_eq!(_res, Ok(slash));
-					Self::deposit_event(Event::<T>::Slashed(current_round, loser.clone(), slash));
-
-					// inform the verifier that they can now try again, if we're still in the signed
-					// validation phase.
-					if crate::Pallet::<T>::current_phase().is_signed_validation() &&
-						Submissions::<T>::has_leader(current_round)
-					{
-						// defensive: verifier just reported back a result, it must be in clear
-						// state.
-						let _ = <T::Verifier as AsynchronousVerifier>::start().defensive();
-					}
-				}
+				Self::handle_solution_rejection(current_round, "Rejected");
 			},
-			VerificationResult::DataUnavailable => {
-				unreachable!("TODO")
+			VerificationResult::DataUnavailable(info) => {
+				sublog!(
+					trace,
+					"signed",
+					"Verification data unavailable for round {}: {:?}",
+					current_round,
+					info
+				);
+
+				match info {
+					DataUnavailableInfo::Page(_) => {
+						// This should not normally happen with current logic - missing pages
+						// should be treated as empty and continue normally to Queued/Rejected.
+						sublog!(
+							error,
+							"signed",
+							"Unexpected VerificationDataUnavailable for page - this indicates a bug"
+						);
+						defensive!("VerificationDataUnavailable for page should never happen");
+					},
+					DataUnavailableInfo::Score => {
+						// Score missing at finalization - reject the solution
+						sublog!(
+							trace,
+							"signed",
+							"Score was unavailable during verification finalization for round {}, rejecting solution",
+							current_round
+						);
+						Self::handle_solution_rejection(
+							current_round,
+							"VerificationDataUnavailable for score",
+						);
+					},
+				}
 			},
 		}
 	}
 }
 
+impl<T: Config> Pallet<T> {
```

Comment:
A scenario that we don't have good tests for AFAIK:

* imagine the duration of the signed phase is such that we can verify half a solution.
* Then we go into `AreWeDone`, and since nothing is queued, we revert back to signed phase.
* No one submits anything.